pyro/contrib/funsor/enum_messenger.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2017-2019 Uber Technologies, Inc.
+# SPDX-License-Identifier: Apache-2.0
+
+import torch
+from collections import OrderedDict
+
+from pyro.poutine import Messenger
+from pyro.poutine.trace_messenger import TraceMessenger
+
+from pyro.contrib.funsor import named, to_funsor, to_data
+
+
+class SimpleEnumMessenger(Messenger):
+
+    def _pyro_sample(self, msg):
+
+        import funsor
+
+        if msg["done"] or msg["is_observed"] or msg.get("expand", False) or \
+                msg["infer"].get("enumerate") != "parallel":
+            return
+
+        msg["infer"]["funsor_fn"] = to_funsor(msg["fn"])
+        size = msg["infer"]["funsor_fn"].inputs["value"].dtype
+        msg["infer"]["funsor_value"] = funsor.Tensor(
+            torch.arange(size),
+            OrderedDict([(msg["name"], size)]),
+            funsor.bint(size)
+        )
+
+        msg["value"] = to_data(msg["infer"]["funsor_value"])
+        msg["done"] = True
+
+
+class FunsorTraceMessenger(TraceMessenger):
+
+    def _pyro_post_sample(self, msg):
+        if "funsor_fn" not in msg["infer"]:
+            msg["infer"]["funsor_fn"] = to_funsor(msg["fn"])
```

Comment:
Compare the use of `to_funsor` during model execution to convert a positional `TorchDistribution` to a named `Funsor` here to the way we currently use `Trace.pack_tensors()` and `pyro.ops.packed` in our ELBO implementations to convert `log_prob` tensors to a packed representation all at once after the model has finished executing, which forces `pyro.poutine.EnumMessenger` and `pyro.poutine.trace` to track and accumulate lots of redundant and otherwise mysterious information about how to interpret batch dimensions.